### PR TITLE
Csp 22 where standard delivered

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/ProviderCourseLocations/ProviderCourseLocationViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        public bool? HasBlockReleaseDeliveryOption { get; set; }
+
+        public string RegionName { get; set; }
+
+        public CourseLocationDeliveryOption DeliveryOption()
+        {
+            if ((HasDayReleaseDeliveryOption.HasValue && HasDayReleaseDeliveryOption.Value) &&
+                (HasBlockReleaseDeliveryOption.HasValue && HasBlockReleaseDeliveryOption.Value))
+            {
+                return CourseLocationDeliveryOption.DayAndBlockRelease;
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Web/Models/ProviderCourseLocations/ProviderCourseLocationViewModel.cs:21]: Enums are not required here. Add a string field in the view model and populate that from the constants.

Comment (MarkFCain) [reply src/SFA.DAS.Roatp.CourseManagement.Web/Models/ProviderCourseLocations/ProviderCourseLocationViewModel.cs:20]: enums removed and replaced with string text

File: src/SFA.DAS.Roatp.CourseManagement.Application/Standard/Queries/GetStandardDetailsQueryHandler.cs [hunk 2]
```diff
@@ -18,18 +18,17 @@ public GetStandardDetailsQueryHandler(IApiClient apiClient, ILogger<GetStandardD
...
         {
             _logger.LogInformation("Get Standards details request received for ukprn {ukprn} and larsCode {larsCode}", request.Ukprn, request.LarsCode);
-            var url = $"ProviderCourse/{request.Ukprn}/Course/{request.LarsCode}";
-                var standardDetails = await _apiClient.Get<Domain.ApiModels.StandardDetails>(url);
-                if (standardDetails == null)
-                {
-                    _logger.LogError("Standard details not found for ukprn {request.Ukprn} and LarsCode {request.LarsCode}");
-                    return null;
-                }
+            var standardDetails = await _apiClient.Get<Domain.ApiModels.StandardDetails>($"ProviderCourse/{request.Ukprn}/Course/{request.LarsCode}");
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Application/Standard/Queries/GetStandardDetailsQueryHandler.cs:21]: That route just looks wrong. It should be providers/{ukprn}/courses/{larscode}

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs [hunk 3]
```diff
@@ -12,11 +15,51 @@ public class StandardDetailsViewModel
...
+
+        public List<string> Regions()
+        {
+            return SubRegionCourseLocations
+                .OrderBy(x => x.RegionName)
+                .Distinct().Select(region => region.RegionName).Distinct().ToList();
+        }
+
+
+        public string LocationSummary()
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs:39]: Can be a get property

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs [hunk 4]
```diff
@@ -12,11 +15,51 @@ public class StandardDetailsViewModel
...
         public string ContactUsPhoneNumber { get; set; }
         public string ContactUsEmail { get; set; }
         public string ContactUsPageUrl { get; set; }
+        public List<ProviderCourseLocationViewModel> ProviderCourseLocations { get; set; }
+
+        public List<ProviderCourseLocationViewModel> SubRegionCourseLocations { get; set; }
+       
+        public ProviderCourseLocationViewModel NationalCourseLocation { get; set; }
+
+        public List<string> Regions()
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs:31]: Can be a get property

File: src/SFA.DAS.Roatp.CourseManagement.Web/Views/Standards/ViewStandardDetails.cshtml [hunk 5]
```diff
@@ -224,4 +225,106 @@
...
+                <table class="govuk-table" aria-describedby="providerCourse-locations">
+                    <thead class="govuk-table__head">
+                    <tr class="govuk-table__row">
+                        <th class="govuk-table__header" scope="col"> Venue name </th>
+                        <th class="govuk-table__header" scope="col"> Delivery method</th>
+                    </tr>
+                    </thead>
+                    <tbody class="govuk-table__body">
+                    @foreach (var location in Model.ProviderCourseLocations)
+                                {
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Web/Views/Standards/ViewStandardDetails.cshtml:261]: alignment

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs [hunk 6]
```diff
@@ -12,11 +15,51 @@ public class StandardDetailsViewModel
...
+        {
+            return SubRegionCourseLocations
+                .OrderBy(x => x.RegionName)
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs:34]: I am unable to follow this linq

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/ProviderCourseLocations/ProviderCourseLocationViewModel.cs (hunk 1 above)
Comment (DevenKShah) [reply src/SFA.DAS.Roatp.CourseManagement.Web/Models/ProviderCourseLocations/ProviderCourseLocationViewModel.cs:20]: This is fine. But I think we will need to move this logic centrally so it becomes more reusable. don't have to do it now, but consider it when we come to next use of it.

File: src/SFA.DAS.Roatp.CourseManagement.Web.UnitTests/Models/Standards/StandardDetailsViewModelTests.cs [hunk 7]
```diff
@@ -41,6 +50,159 @@ public void ImplicitOperator_ConvertsFromStandardDetails(string regulatorName, b
...
             viewModel.IsStandardRegulated.Should().Be(isRegulated);
             viewModel.CourseDisplayName.Should().Be(expectedCourseDisplayName);
+            viewModel.StandardInfoUrl.Should().Be(standardInfoUrl);
+            viewModel.ContactUsEmail.Should().Be(contactUsEmail);
+            viewModel.ContactUsPhoneNumber.Should().Be(contactUsPhoneNumber);
+            viewModel.ContactUsPageUrl.Should().Be(contactUsPageUrl);
+        }
+
+        [Test]
+        public void ImplicitOperator_ConvertsCourseLocationsFromStandardDetails()
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Web.UnitTests/Models/Standards/StandardDetailsViewModelTests.cs:60]: This is missing tests for Regions and LocationSummary which is calculated in the model. 

If you are adding the above coverage, I suggest you use one test class per method/property.

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs [hunk 8]
```diff
@@ -35,7 +78,21 @@ public class StandardDetailsViewModel
...
                 ContactUsPageUrl = standardDetails.ContactUsPageUrl,
+                ProviderCourseLocations = standardDetails.ProviderCourseLocations.Where(a => a.LocationType == LocationType.Provider).Select(x => (ProviderCourseLocationViewModel)x).ToList(),
+                SubRegionCourseLocations = standardDetails.ProviderCourseLocations.Where(a => a.LocationType == LocationType.Regional).Select(x => (ProviderCourseLocationViewModel)x).ToList(),
+                NationalCourseLocation = standardDetails.ProviderCourseLocations.Where(a => a.LocationType == LocationType.National).Select(x => (ProviderCourseLocationViewModel)x).FirstOrDefault()
             };
         }
     }
+
+
+    public static class WhereIsCourseDelivered
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs:89]: CourseDeliveryMessageFor

**Review by DevenKShah — APPROVED** [review]
I have few suggestions, but nothing that should hold up the story any further.

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs:39]: I think SonarCloud reported it as a code smell when I did that, so I changed it to this.  I'll try it, and if SonarCloud complains, I'll put it back again

File: src/SFA.DAS.Roatp.CourseManagement.Web.UnitTests/Models/Standards/StandardDetailsViewModelTests.cs (hunk 7 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.CourseManagement.Web.UnitTests/Models/Standards/StandardDetailsViewModelTests.cs:61]: LocationSummary coverage was there, regions coverage added

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs (hunk 8 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs]: yes, done

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs (hunk 4 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs:31]: I changes to a get, but SonarCloud identifies that as a critical codesmell, and said to convert to a method, so set back again.

File: src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.CourseManagement.Web/Models/Standards/StandardDetailsViewModel.cs]: Yes, this is now done

**Pull request closed without merging**